pkg/controller/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,318 @@
+	c.excluderChanged = true
+}
+
+func (c *CacheManager) AddObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	isNamespaceExcluded, err := c.processExcluder.IsNamespaceExcluded(process.Sync, instance)
+	if err != nil {
+		return fmt.Errorf("error while excluding namespaces: %w", err)
+	}
+
+	// bail because it means we should not be
+	// syncing this gvk
+	if isNamespaceExcluded {
+		c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+		return nil
+	}
+
+	syncKey := syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName())
+	_, err = c.opa.AddData(ctx, instance)
+	if err != nil {
+		c.syncMetricsCache.AddObject(
+			syncKey,
+			syncutil.Tags{
+				Kind:   instance.GetKind(),
+				Status: metrics.ErrorStatus,
+			},
+		)
+
+		return err
+	}
+
+	c.tracker.ForData(instance.GroupVersionKind()).Observe(instance)
+
+	c.syncMetricsCache.AddObject(syncKey, syncutil.Tags{
+		Kind:   instance.GetKind(),
+		Status: metrics.ActiveStatus,
+	})
+	c.syncMetricsCache.AddKind(instance.GetKind())
+
+	return err
+}
+
+func (c *CacheManager) RemoveObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	if _, err := c.opa.RemoveData(ctx, instance); err != nil {
+		return err
+	}
+
+	// only delete from metrics map if the data removal was succcesful
+	c.syncMetricsCache.DeleteObject(syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName()))
+	c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+
+	return nil
+}
+
+func (c *CacheManager) wipeData(ctx context.Context) error {
+	if _, err := c.opa.RemoveData(ctx, target.WipeData()); err != nil {
+		return err
+	}
+
+	// reset sync cache before sending the metric
+	c.syncMetricsCache.ResetCache()
+	c.syncMetricsCache.ReportSync()
+
+	return nil
+}
+
+func (c *CacheManager) ReportSyncMetrics() {
+	c.syncMetricsCache.ReportSync()
+}
+
+func (c *CacheManager) listAndSyncDataForGVK(ctx context.Context, gvk schema.GroupVersionKind) error {
+	u := &unstructured.UnstructuredList{}
+	u.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind + "List",
+	})
+
+	err := c.reader.List(ctx, u)
+	if err != nil {
+		return fmt.Errorf("replaying data for %+v: %w", gvk, err)
+	}
+
+	defer c.ReportSyncMetrics()
+
+	for i := range u.Items {
+		if err := c.AddObject(ctx, &u.Items[i]); err != nil {
+			return fmt.Errorf("adding data for %+v: %w", gvk, err)
+		}
+	}
+
+	return nil
+}
+
+func (c *CacheManager) updateDatastore(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-c.replayTicker.C:
+			// snapshot the current spec so we can make a step upgrade
+			// to the contests of the opa cache.
+			c.mu.RLock()
+			currentGVKsInAgg := watch.NewSet()
+			currentGVKsInAgg.Add(c.gvkAggregator.ListAllGVKs()...)
+			excluderChanged := c.excluderChanged
+			c.mu.RUnlock()
+
+			c.makeUpdatesForSpecInTime(ctx, currentGVKsInAgg, excluderChanged)
```

Comment:
This would need to be part of the mutex

Reply:
Since it sets/unsets excluderChanged

Reply:
Actually, it's probably fine, but easier to go with the simpler mutex arrangement and make it more complex as necessary... controllers are background processes so a bit of extra latency for set calculations is not an issue. Replay should happen async, though

Reply:
so the crux of snapshotting the spec (read gvks in aggregator and excluder) is to decouple A) adding sources that specify which gvks to add/ remove and the process excluder changes from B) opa cache management. In particular, the calls to the k8s server (list) will be taken out of the online path so they don't block the path. 

---

other things to note:

> Since it unsets excluderChanged

I think I may have a way to not have to modify the spec

Reply:
1) We should not have a tick that delays syncing net-new GVKs

2) There are ways of handling updates to the intent that do not require re-listing to hold a lock (it should be run on a separate thread)

3) Trying to be clever about juggling locks complicates reasoning about data consistency. It may be necessary, but that should be empirically proven (i.e. we shouldn't prematurely optimize).

To be clear, the structure of this code is likely to fundamentally change given it doesn't support watching ASAP, nor does it support retries in the listing. Fixing that will likely moot this conversation.

Reply:
We watch net new gvks in `AddSource` now.